enhancements/console/customize-perspectives.md

```diff
@@ -0,0 +1,294 @@
+
+OpenShift Container Platform (OCP) web console provides two perspectives: Administrator and Developer. 
+The Developer perspective provides workflows specific to developer use cases 
+like create, deploy and monitor applications, while Administrator perspective 
+is responsible for managing the cluster resources, users, and projects.
+
+This document provides details about how cluster administrator 
+can hide user perspectives. 
+
+![Perspectives switch dropdown](perspectives-switch.png)
+
+## Motivation
+
+Cluster administrator wants to hide the "Administrator" perspective 
+from non-privileged and to hide the "Developer" 
+perspective for all users.
+
+### User Stories
+
+#### Story 1
+
+As a cluster admin, I want to hide "Administrator" perspective 
+from all non-privileged users so that non-privileged users
+cannot access "Administrator" perspective for managing the cluster resources, 
+users, and projects. Non-privileged users are users that cannot get
+all `namespaces` (based on RBAC check)
+
+#### Story 2
+
+As a cluster admin, I want to hide "Developer" perspective
+for all users.
+
+#### Story 3
+
+As a cluster admin, I want to hide any perspective contributed by a plugin.
+
+Additional perspectives can be dynamically added with the 
+[Perspective extension](https://github.com/openshift/console/blob/master/frontend/packages/console-dynamic-plugin-sdk/src/extensions/perspectives.ts).
+
+### Goals
+
+- Allow cluster admins to hide the user perspectives.
+- Provide a way that cluster admins can do this based on resource permissions (RBAC).
+
+### Non-Goals
+
+- Allow cluster admins to customize the navigation menu along with perspectives for users or groups.
+
+## Proposal
+
+We can extend the existing `operator.openshift.io/v1` / `Console` CRD. 
+We can add 
+`spec.customization.perspectives` which can have list of customization details.
+
+### Workflow Description
+
+To disable perspective for users with limited permissions we need a list of options with a perspective id and the required permissions.
+
+```text
+├── Spec
+│   ├── customization
+|       ├── perspectives 
+|           ├── id
+|           ├── requiresAccessReview
+|               ├── kind
+|               ├── verb
+└── ...
+```
+
+This might be extendable in the future if we want replace (hide one, show another) a perspective based on a resource permissions:
+
+```text
+├── Spec
+│   ├── customization
+|       ├── perspectives 
+|           ├── id
+|           ├── requiresAccessReview
+|           |   ├── kind
+|           |   ├── verb
+|           ├── missingAccessReview
+|               ├── kind
+|               ├── verb
+└── ...
+```
+
+Each perspective conforms to the following schema:
+
+- `id: string` , defines the id of perspective to disable. Incorrect or unknown ids will be ignored.
+
+// +optional
+- `requiresAccessReview: authV1.ResourceAttributes[]` , defines a list of permission checks. The perspective will only be shown when all checks are successful.
```

Comment:
Would be good to explain the use case for this and missing

It seems to me they are, show the perspective only when a use has all of these permissions, and then show the perspective whenever they have some of a list of permissions? 

How do the two interact together?

Reply:
Hey, yes, both can work together esp. in the case where you want to show **another** perspective to users **without** specific permissions.

For example, a plugin or customer adds a new "Monitoring" perspective, which works only with specific permissions.

The admin can then define that this perspective is shown with some `requiresAccessReview` permissions.

But to show another **limited** "Monitoring" (same name, different ID!) perspective to all the others, the admin can define a `missingAccessReview` check so that the users without access to specific resources see it - but the users with access to the first "Monitoring" perspective doesn't see both.

Reply:
OK that makes sense, please make sure to explain this within the enhancement!

Do you see a need to have a concept of `anyOf` and `allOf` within the access review? Allowing people to define, I want to show this perspective if they have all of these, or if they have any of these. Likewise with missing?

Reply:
Added in the document. 

For `anyOf` and `allOf`, it is a good idea to consider but we are not adding that for now. We will make use of `requiresAccessReview` and `missingAccessReview`.

Reply:
Just to be clear, I'm assuming some conversation happened here. Is this something you might want to add in the future? If it's something you may want in the future we should design the API in a way that it's compatible to add this later, however, if it's something you never want, we don't need to consider it now.